x86_toolchain.1
.TH X86_TOOLCHAIN.SH 1 25-10-2023

.SH NAME

x86_toolchain.sh \- x86-64 based compiler Toolchain

.SH SYNOPSIS

.B x86_toolchain.sh
[\fBoptions\fR]
<\fIassembly_filename\fR>
[\fB\-o | \-\-output\fR <\fIoutput_filename\fR>]

.SH DESCRIPTION
.ad l 
.nh
This toolchain was built for applying modality and ease of use to compiling and debugging NASM based assembly code.
It uses \fBnasm\fR to compile and \fBld\fR to link with options to debug and run inside \fBgdb\fR.
This version was modified to run in x86-64 bit architecture by default.

The goal of this project is to create a tool that simplifies the process of compilation and debugging by reducing the number of steps and instructions required. 

Instructions for toolchain installation can be located at: \fIhttps://github.com/LubosKuzma/ITSC204/blob/main/scripts/README.md

.SH OPTIONS

.TP
.BR \-v ", " \-\-verbose                
	Show more information about steps performed and the set arguments.
.TP
.BR \-g ", " \-\-gdb                    
	Run gdb command on executable.
.TP
.BR \-b ", " "\-\-break " <\fIbreakpoint\fR>    
	Add breakpoint after running \fBgdb\fR.
.TP
.BR \-r ", " \-\-run                    
	Run program in \fBgdb\fR automatically. Same as run command inside \fBgdb\fR environment.
.TP
.BR \-q ", " \-\-qemu                   
	Run executable in \fBqemu\fR emulator. This will execute the program.
.TP
.BR \-32 ", " \-\-x86\-32              
	Compile for 32-bit (x86-32) system.
.TP
.BR \-o ", " "\-\-output " <\fIfilename\fR>      
	Desired name for new output file.
	
.SS GDB / GEF specific commands:
After compilation and executing the file using \fGDB\fR, the user is able to control the debugger as normal.
GDB offers extensive functionality in debugging. Below is a list of some of the more common ones:
.TP
\fBb\fR - Puts a breakpoint at the current line
.PP
	Alternate usage: \fBb\fR [\fBoption\fR]

	[\fBmain\fR] - Puts a breakpoint at the beginning of the program

	[\fBN\fR] - Puts a breakpoint at line N

	[\fB+N\fR] - Puts a breakpoint N lines down from the current line

	[\fBn\fR] - Puts a breakpoint at the beginning of function "fn"

	[\fBN\fR] - Deletes breakpoint number N

\fBinfo break\fR - list breakpoints

\fBr\fR - Runs the program until a breakpoint or error

\fBc\fR - Continues running the program until the next breakpoint or error

\fBf\fR - Runs until the current function is finished

\fBs\fR - Runs the next line of the program

	Alternate usage: \fBs\fR [\fBoption\fR]

	[\fBN\fR] - Runs the next N lines of the program

\fBn\fR - Like s, but it does not step into functions

\fBp\fR <\fIvar\fR> - Prints the current value of the variable "var"

\fBbt\fR - Prints a stack trace

\fBu\fR - Goes up a level in the stack

	Alternate usage: \fBu\fR [\fBoption\fR]

	[\fBN\fR] - Runs until you get N lines in front of the current line

\fBd\fR - Goes down a level in the stack

\fBq\fR - Quits gdb
	
.SH PREREQUISITES

.SS x86-64 Bit architecture:
.BR gcc ", " gdb ", " gef ", " qemu

.SS Arm architecture:
.BR arm\-gcc ", "multiarch ", " gdb ", " gef ", " qemu

.SS Add toolchain script to PATH:
.TP
.ad l
\fBFor bash:\fR
.TP
"echo \(aqexport PATH=$PATH:<\fIfilename\fR>\(aq >>/.bashrc"
.TP 
\fBFor zsh:\fR
.TP
"echo \(aqexport PATH=$PATH:<\fIfilename\fR>\(aq >>/.zsh"

.SH BUGS

No known bugs. 
.sp 0
If you find a bug, please report it at \fIhttps://github.com/LubosKuzma

.SH SEE ALSO

\fBgdb\fR(1), \fBnasm\fR(1), \fBld\fR(1)

.SH AUTHOR
.TP
Lubos Kuzma \fIhttps://www.linkedin.com/in/lubos-kuzma-0719a586
ISS program, SADT, SAIT. August 2022

x86_toolchain.sh

#!/bin/bash

# Script information
# Created by Lubos Kuzma
# ISS Program, SADT, SAIT
# August 2022
# Edited By Brian Dang
# Default setting to x64 Architecture
# Added comments explaining each section of the script to enhance readability and understanding.
# Try to add an <Install> option but it seem to not recognize -i.

# Check if there are not enough arguments provided
if [ $# -lt 1 ]; then
    echo "Usage:"
    echo ""
    echo "x86_toolchain.sh [ options ] <assembly filename> [-o | --output <output filename>]"
    echo ""
    echo "-v | --verbose                Show some information about steps performed."
    echo "-g | --gdb                    Run gdb command on executable."
    echo "-b | --break <break point>    Add breakpoint after running gdb. Default is _start."
    echo "-r | --run                    Run program in gdb automatically. Same as run command inside gdb env."
    echo "-q | --qemu                   Run executable in QEMU emulator. This will execute the program."
    echo "-o | --output <filename>      Output filename."
    #echo "-i | --install                install prerequisite for toolchain."
    exit 1
fi

# Initialize variables
POSITIONAL_ARGS=()
GDB=False
OUTPUT_FILE=""
VERBOSE=False
QEMU=False
BREAK="_start"
RUN=False
INSTALL=False

# Parse command-line arguments
while [[ $# -gt 0 ]]; do
    case $1 in
        -g|--gdb)
            GDB=True
            shift # past argument
            ;;
        -o|--output)
            OUTPUT_FILE="$2"
            shift # past argument
            shift # past value
            ;;
        -v|--verbose)
            VERBOSE=True
            shift # past argument
            ;;

        -q|--qemu)
            QEMU=True
            shift # past argument
            ;;
        -r|--run)
            RUN=True
            shift # past argument
            ;;
        -b|--break)
            BREAK="$2"
            shift # past argument
            shift # past value
            ;;
        -*|--*)
            echo "Unknown option $1"
            exit 1
            ;;
        *)
            POSITIONAL_ARGS+=("$1")
            shift # past argument
            ;;
 #       -i|--install)
 #           INSTALL=True
 #           shift # past argument
 #           ;;
    esac
done

# Set positional arguments
set -- "${POSITIONAL_ARGS[@]}"

# Check if the specified file exists
if [[ ! -f $1 ]]; then
    echo "Specified file does not exist"
    exit 1
fi

# Set the output file if not provided
if [ "$OUTPUT_FILE" == "" ]; then
    OUTPUT_FILE=${1%.*}
fi

# Display information if verbose mode is enabled
if [ "$VERBOSE" == "True" ]; then
    echo "Arguments being set:"
    echo "    GDB = ${GDB}"
    echo "    RUN = ${RUN}"
    echo "    BREAK = ${BREAK}"
    echo "    QEMU = ${QEMU}"
    echo "    Input File = $1"
    echo "    Output File = $OUTPUT_FILE"
    echo "    Verbose = $VERBOSE"
    echo "    64 bit mode = $BITS"
    echo ""
    echo "NASM started..."
fi

# Compile the assembly file using NASM
nasm -f elf64 $1 -o $OUTPUT_FILE.o && echo ""

# Display information if verbose mode is enabled
if [ "$VERBOSE" == "True" ]; then
    echo "NASM finished"
    echo "Linking ..."
fi

# Link the object file using the ld linker
ld -m elf_x86_64 $OUTPUT_FILE.o -o $OUTPUT_FILE && echo ""

# Display information if verbose mode is enabled
if [ "$VERBOSE" == "True" ]; then
    echo "Linking finished"
fi

# Run the executable in QEMU if specified
if [ "$QEMU" == "True" ]; then
    echo "Starting QEMU ..."
    echo ""
    qemu-x86_64 $OUTPUT_FILE && echo ""
    exit 0
fi

# Run the GDB debugger if specified
if [ "$GDB" == "True" ]; then
    gdb_params=()
    gdb_params+=(-ex "b ${BREAK}")
    if [ "$RUN" == "True" ]; then
        gdb_params+=(-ex "r")
    fi
    gdb "${gdb_params[@]}" $OUTPUT_FILE
fi
# Run installation script
#if [ "$INSTALL" == "True" ]; then
 #   echo "Starting Prerequisite install..."
  #  bash x86_toolchain_installer.sh
#fi